src/main/java/org/opensearch/flowframework/transport/DeprovisionWorkflowTransportAction.java

```diff
@@ -0,0 +1,335 @@
+        List<ProcessNode> provisionProcessSequence,
+        ActionListener<WorkflowResponse> listener
+    ) {
+        // Create a list of ProcessNodes with ta corresponding deprovision workflow steps
+        List<ProcessNode> deprovisionProcessSequence = provisionProcessSequence.stream()
+            // Only include nodes that created a resource
+            .filter(pn -> resourceMap.containsKey(pn.id()))
+            // Create a new ProcessNode with a deprovision step
+            .map(pn -> {
+                String stepName = pn.workflowStep().getName();
+                String deprovisionStep = getDeprovisionStepByWorkflowStep(stepName);
+                // Unimplemented steps presently return null, so skip
+                if (deprovisionStep == null) {
+                    return null;
+                }
+                // New ID is old ID with deprovision added
+                String deprovisionStepId = pn.id() + DEPROVISION_SUFFIX;
+                return new ProcessNode(
+                    deprovisionStepId,
+                    workflowStepFactory.createStep(deprovisionStep),
+                    Collections.emptyMap(),
+                    new WorkflowData(
+                        Map.of(getResourceByWorkflowStep(stepName), resourceMap.get(pn.id()).resourceId()),
+                        workflowId,
+                        deprovisionStepId
+                    ),
+                    Collections.emptyList(),
+                    this.threadPool,
+                    pn.nodeTimeout()
+                );
+            })
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        // Deprovision in reverse order of provisioning to minimize risk of dependencies
+        Collections.reverse(deprovisionProcessSequence);
+        logger.info("Deprovisioning steps: {}", deprovisionProcessSequence.stream().map(ProcessNode::id).collect(Collectors.joining(", ")));
+
+        // Repeat attempting to delete resources as long as at least one is successful
+        int resourceCount = deprovisionProcessSequence.size();
+        while (resourceCount > 0) {
+            Iterator<ProcessNode> iter = deprovisionProcessSequence.iterator();
+            while (iter.hasNext()) {
+                ProcessNode deprovisionNode = iter.next();
+                ResourceCreated resource = getResourceFromDeprovisionNode(deprovisionNode, resourceMap);
+                String resourceNameAndId = getResourceNameAndId(resource);
+                CompletableFuture<WorkflowData> deprovisionFuture = deprovisionNode.execute();
+                try {
+                    deprovisionFuture.join();
+                    logger.info("Successful {} for {}", deprovisionNode.id(), resourceNameAndId);
+                    // Remove from list so we don't try again
+                    iter.remove();
+                    // Pause briefly before next step
+                    Thread.sleep(100);
```

Comment:
Curious, why do we need to pause here?

Reply:
Great question that I've asked myself multiple times.

The sequence I've been testing with is "undeploy model" followed by "delete model" followed by "delete connector".

With no delay, the "undeploy" step returns a successful result but attempts to delete a model immediately following sometimes show it's still deployed.  It probably only needs a few milliseconds delay and 100 is excessive. But also short.

There's a similar issue with "delete" model prior to deleting the connector. This seems to have a longer delay over 500ms but always less than 1s.  I've handled this in the "repeat the whole workflow" logic.

Reply:
Should we give some variable time here, on big clusters where the model is deployed to 30+nodes, I wonder if we will encounter some issue here

Reply:
I don't think undeploying takes that much time (the responses all return DELETED) so this is just index replication time, which should always be under 1 second, but I don't want a 1 second delay between steps.

Worst case it doesn't delete and they can call it again.

Reply:
@amitgalitz let me know if we can merge this as is or if you want me to add settings to control the time delays.

Reply:
I am okay merging as is, if it breaks we will fix it :)